src/datasets/utils/py_utils.py

```diff
@@ -356,6 +385,26 @@ def memoize(self, obj):
         if type(obj) != str:
             dill.Pickler.memoize(self, obj)
 
+    @staticmethod
+    def maybe_register_superfunc(obj):
+        """Dynamically registers a custom pickling function for the type of this object `obj`
+        by checking the `Pickler.sublcass_dispatch` table. All super-classes of `obj` are
+        traversed through mro() and checked against the table. If there is a match between
+        a super-class and the table, then we add the type() of the object along with the function
+        of the found super-class to Pickler.dispatch. This will internally tell Pickle to use
+        this function, originally related to the super-class, for this function as well."""
+        t = type(obj)
+        if t not in Pickler.dispatch and t not in Pickler.sublcass_dispatch:
+            if hasattr(obj.__class__, "__mro__"):
+                for supercls in obj.__class__.__mro__:
+                    if supercls in Pickler.sublcass_dispatch:
+                        pklregister(t)(Pickler.sublcass_dispatch[supercls])
+                        # Break as soon as one match is found. We only want to most direct
+                        # ancestor class. _Some_ caution is needed here because the order
+                        # that we traverse is that of MRO, which is depth-first with some quirks
+                        # see # https://stackoverflow.com/a/2010732/1150683
+                        break
```

Comment:
@mariosasko  Absolutely right about __mro__. I have a (bad) habit of checking functionality for _old_ versions to be sure but it isn't needed here.

About the classmethod: I don't think so. As far as I understand, `save` is called on an instance (not a class), which will then call `maybe_register_superfunc`. So `maybe_register_superfunc` should only ever be called by an instance, never by the Class itself.

Reply:
Oh, and we can change `Pickler.sublcass_dispatch` to `self.sublcass_dispatch` if you prefer. If no instance attribute exists with the name, the class variable will be used, as is our case.

Reply:
My only concern is subclassing, which requires either `classmethod` or replacing `Pickler` with `self` as you suggest. IMO `classmethod` is cleaner because it avoids possible attribute clash and communicates intent better. 

> As far as I understand, save is called on an instance (not a class), which will then call maybe_register_superfunc. So maybe_register_superfunc should only ever be called by an instance, never by the Class itself

But the same case could also be made for `staticmethod` because it's bound to a class, and not the instance.

Reply:
I guess it's a matter of semantics/preference. If you prefer a class method, that's fine.